testing/test_windows_subst.py

```diff
@@ -0,0 +1,95 @@
+import os.path
+import subprocess
+import sys
+from string import ascii_lowercase
+
+from py._path.local import LocalPath
+
+import pytest
+from _pytest import pytester
+from _pytest.fixtures import FixtureRequest
+
+
+class _SubstTmpdirFactory:
+    """
+    A temp dir factory which may do a subst on Windows drives.
+    """
+
+    def __init__(self, tmpdir, tmpdir_factory, subst):
+        self.tmpdir = tmpdir
+        self.tmpdir_factory = tmpdir_factory
+        self.subst = subst
+        self.subst_drives = {}
+
+    def _subst_path(self, filename):
+
+        if self.subst:
+            for c in ascii_lowercase[7:]:  # Create a subst drive from H-Z.
+                c += ":"
+                if not os.path.exists(c):
+                    drive = c
+                    break
+            else:
+                raise AssertionError("Unable to find suitable drive letter for subst.")
+
+            directory = os.path.dirname(filename)
+            basename = os.path.basename(filename)
+
+            args = ["subst", drive, directory]
+            subprocess.check_call(args)
+            assert os.path.exists(drive)
+            self.subst_drives[drive] = directory
+
+            filename = LocalPath(os.path.join(drive, basename))
+
+        return filename
+
+    def mktemp(self, name, numbered):
+        filename = self.tmpdir_factory.mktemp(name, numbered=numbered)
+        filename = self._subst_path(filename)
+        return filename
+
+    def unsubst(self):
+        for drive in self.subst_drives:
+            args = ["subst", "/D", drive]
+            subprocess.check_call(args)
+
+
+@pytest.fixture(params=[{"subst": True}, {"subst": False}])
+def _subst_tmpdir_factory(request: FixtureRequest, tmpdir, tmpdir_factory):
+    if sys.platform != "win32":
+        pytest.skip("Windows only test.")
+    factory = _SubstTmpdirFactory(tmpdir, tmpdir_factory, **request.param)
+    yield factory
+    factory.unsubst()
+
+
+@pytest.fixture
+def _custom_testdir(
+    request: FixtureRequest, _subst_tmpdir_factory
+) -> "pytester.Testdir":
+    return pytester.Testdir(request, _subst_tmpdir_factory)
+
+
+def test_windows_subst_resolve(_custom_testdir, _subst_tmpdir_factory) -> None:
```

Comment:
Looking at the test, seems like it is a bit of overkill to override testdir and tmpdir_factory... I wonder if it is possible to just have a context manager instead:

```python

@contextmanager
def subst_path(path):
    # make path.parent subst to another drive
    try:
        yield subs_path
    finally:
        # undo subst

```

Then use in the test like this:

```python
@pytest.mark.skipif(not sys.platform.startswith("win"), reason="Windows only")
def test_windows_subst_resolve(subst, testdir) -> None:
    """
    Check that when we have a subst on Windows the errors reported are not
    in the subst target (See #5965).
    """
    p = testdir.makepyfile(
        """
        import pytest
        def test_foo():
            raise AssertionError()
        """
    )    
    with subst_path(p) as subst_p: 
        result = testdir.runpytest(subst_p, "-v")

        # Make sure that the error is reported as the 'subst' path, not as an
        # absolute (resolved) path.        
        result.stdout.fnmatch_lines([
            str(subst_path) + '*',
        ])        
```

Reply:
Done